### PR TITLE
MAIN: Enable support for Sphinx>=4

### LINKED ISSUE #338: Allow using Sphinx 4
# Summary

Right now `sphinx>=2,<4` in [setup.cfg](https://github.com/executablebooks/MyST-NB/blob/e39132e4f1c58352a9f6d3004c7336dbbeca4e8a/setup.cfg#L51) makes it impossible to use `myst-nb` with Sphinx 4. `myst-parser` upgraded to Sphinx 4 recently https://github.com/executablebooks/MyST-Parser/pull/390.

# Tasks to complete

- [ ] Try removing the Sphinx to `<5` here
      https://github.com/executablebooks/MyST-NB/blob/master/setup.cfg#L51
      and bump myst-parser to the latest version here:
      https://github.com/executablebooks/MyST-NB/blob/master/setup.cfg#L47
- [ ] Run tests and make changes needed to make them pass
- [ ] Merge in, then we should cut a new release

### PR DESCRIPTION
**Pull request by mmcky**
This PR enables support for `sphinx>=4` and drops support for `sphinx<=2`

fixes #338

# Tasks

- [x] rework tests to specify `sphinx3` and `sphinx4` for tests that differ (rather than all test cases)
- [x] Wait for #353 to be merged (which should actually be 0.15.2)
- [x] Rebase onto this branch
- [x] Fix changes needed just for Sphinx 4

This work has all been migrated to #356 so will close this PR. Thanks @chrisjsewell @choldgraf for comments and integration points with `myst-parser`

### CODE REVIEW

**Comment by mmcky** [comment]
@chrisjsewell the current commit to this should setup a test environment through tox that includes `sphinx4` but it doesn't seem to work for me. It keeps installing `sphinx=3.5.4`. Any ideas?

I widened the `sphinx` version requirement in `setup.cfg` and then added a `sphinx4` deps section in the `tox.ini`

**Comment by chrisjsewell** [comment]
myst-parser 0.14 does not support sphinx v4:https://github.com/executablebooks/MyST-Parser/blob/master/CHANGELOG.md#0150---2021-06-13

**Comment by mmcky** [comment]
> myst-parser 0.14 does not support sphinx v4:https://github.com/executablebooks/MyST-Parser/blob/master/CHANGELOG.md#0150---2021-06-13

Ah of course. Thanks @chrisjsewell. I'll pin to `myst-parser=0.15`. Thank you!

**Comment by chrisjsewell** [comment]
Pin `myst-parser~=0.15.1` 👍

**Comment by chrisjsewell** [comment]
Note: when upgrading jupyter-book with this new version, it will be going from myst-parser v0.13 -> v0.15.
There are a number of additions to configurations to consider: https://github.com/executablebooks/MyST-Parser/blob/master/CHANGELOG.md#0150---2021-06-13

- Adding tasklist to the default `myst_enable_extensions` will close https://github.com/executablebooks/jupyter-book/issues/1133
- setting `myst_dmath_double_inline=True` will close https://github.com/executablebooks/jupyter-book/issues/1181
   - perhaps that could also be set to true by default here (currently set to False, to be overly cautious about being back-compatible)
- the `myst_url_schemes` can be removed from the default config, since it is now set already
- sphinx 4 will close https://github.com/executablebooks/jupyter-book/issues/963, but at the same time user who were previously setting `mathjax_config` should be warned that they now need to set `mathjax2_config`/`mathjax3_config` https://www.sphinx-doc.org/en/master/usage/extensions/math.html#confval-mathjax3_config
- users should be warned about the removed deprecations: https://github.com/executablebooks/MyST-Parser/blob/master/CHANGELOG.md#remove-v013-deprecations-%EF%B8%8F

**Comment by mmcky** [comment]
Great points from @chrisjsewell:

- [ ] test fixtures need to track `docutils` https://github.com/executablebooks/MyST-Parser/blob/be4838c35d374a51427a92a2736b29a9abc018d4/setup.cfg#L74
- [ ] test fixtures need to track `sphinx` https://github.com/executablebooks/MyST-Parser/blob/be4838c35d374a51427a92a2736b29a9abc018d4/tests/test_sphinx/test_sphinx_builds.py#L101

**Comment by mmcky** [comment]
So `myst-parser==0.15.1` seems responsible for the changes in `fixture` outputs. Maybe this is due to the `docutils==0.17` rather than a `sphinx3` vs `sphinx4` update. @chrisjsewell from your work on `myst-parser` tracking `sphinx` versions is still going to be needed going forward right?

If you look at diff for: `tests/test_parser/test_complex_outputs.sphinx3.xml`

those outputs were built with `sphinx3` but `the `tox` environment moved up to `myst-parser=0.15.1` and `docutils~=0.16`

File: setup.cfg [hunk 1]
```diff
@@ -38,17 +38,17 @@ project_urls =
...
     ipywidgets>=7.0.0,<8
     jupyter-cache~=0.4.1
     jupyter_sphinx~=0.3.2
-    myst-parser~=0.14.0
+    myst-parser~=0.15.1
     nbconvert>=5.6,<7
     nbformat~=5.0
     pyyaml
-    sphinx>=2,<4
+    sphinx>=2,<5
```
Comment (chrisjsewell) [inline setup.cfg:51]: here we should also drop sphinx v2 support, in myst-parser it is `sphinx>=3.1,<5` (see https://github.com/executablebooks/MyST-Parser/pull/400)

Comment (mmcky) [reply setup.cfg:51]: thanks @chrisjsewell I will copy your pins from `myst-parser` 👍

**Comment by mmcky** [comment]
Closing this as work is now contained in #356

**Pull request closed without merging**